pallets/farming/src/gauge.rs

```diff
@@ -0,0 +1,494 @@
+// Copyright 2020-2023 Manta Network.
+// This file is part of Manta.
+//
+// Manta is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Manta is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Manta.  If not, see <http://www.gnu.org/licenses/>.
+
+use codec::HasCompact;
+use frame_support::pallet_prelude::*;
+use scale_info::TypeInfo;
+use sp_core::U256;
+use sp_runtime::{
+    traits::{Zero, *},
+    ArithmeticError, Perbill, RuntimeDebug, SaturatedConversion,
+};
+use sp_std::prelude::*;
+
+use crate::*;
+
+#[derive(Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub struct GaugeInfo<BalanceOf: HasCompact, BlockNumberFor, AccountIdOf> {
+    pub who: AccountIdOf,
+    pub gauge_amount: BalanceOf,
+    pub total_time_factor: u128,
+    pub latest_time_factor: u128,
+    pub claimed_time_factor: u128,
+    pub gauge_start_block: BlockNumberFor,
+    pub gauge_stop_block: BlockNumberFor,
+    pub gauge_last_block: BlockNumberFor,
+    pub last_claim_block: BlockNumberFor,
+}
+
+impl<BalanceOf, BlockNumberFor, AccountIdOf> GaugeInfo<BalanceOf, BlockNumberFor, AccountIdOf>
+where
+    BalanceOf: Default + HasCompact,
+    BlockNumberFor: Default,
+{
+    fn new(who: AccountIdOf) -> Self {
+        Self {
+            who,
+            gauge_amount: Default::default(),
```

Comment:
I'm not against `Default` in runtime, but yeah, we should be careful about these `Default`.